File: test-api.py
#!/usr/bin/env python3

"""
JSON Canvas AI - Python Client Test Suite
Complete test of all headless API endpoints
Run with: python test-api.py
"""

import requests
import json
import sys

BASE_URL = "http://localhost:9002/api"

def test_endpoint(name, url, method="GET", data=None):
    """Test an API endpoint and print results"""
    print(f"📋 Testing: {name}")
    
    try:
        if method == "GET":
            response = requests.get(url)
        elif method == "POST":
            response = requests.post(url, json=data, headers={"Content-Type": "application/json"})
        elif method == "PUT":
            response = requests.put(url, json=data, headers={"Content-Type": "application/json"})
        elif method == "DELETE":
            response = requests.delete(url)
        
        result = response.json()
        
        if response.status_code == 200:
            print(f"✅ {name} - SUCCESS")
            if "data" in result:
                preview = str(result["data"])[:200] + "..." if len(str(result["data"])) > 200 else str(result["data"])
                print(f"📄 Result: {preview}")
            else:
                preview = str(result)[:200] + "..." if len(str(result)) > 200 else str(result)
                print(f"📄 Response: {preview}")
        else:
            print(f"❌ {name} - FAILED")
            print(f"📄 Error: {result}")
        
        print()
        return result
        
    except Exception as e:
        print(f"❌ {name} - ERROR: {str(e)}")
        print()
        return None

class JSONCanvasClient:
    """Complete Python client for JSON Canvas AI"""
    
    def __init__(self, base_url=BASE_URL):
        self.base_url = base_url
    
    def convert_text_to_json(self, text, instructions=""):
        """Convert text to structured JSON"""
        response = requests.post(f"{self.base_url}/ai/convert-text", json={
            "rawText": text,
            "instructions": instructions
        })
        
        if response.status_code == 200:
            result = response.json()
            if result.get("success"):
                return json.loads(result["data"]["generatedJson"])
        
        raise Exception(f"API Error: {response.text}")
    
    def enhance_field(self, content, prompt):
        """Enhance a field using AI"""
        response = requests.post(f"{self.base_url}/ai/enhance-field", json={
            "fieldContent": content,
            "userPrompt": prompt
        })
        
        if response.status_code == 200:
            result = response.json()
            if result.get("success"):
                return result["data"]["enhancedContent"]
        
        raise Exception(f"API Error: {response.text}")
    
    def format_json(self, json_string, instructions=""):
        """Format and fix JSON"""
        response = requests.post(f"{self.base_url}/ai/format-json", json={
            "jsonString": json_string,
            "instructions": instructions
        })
        
        if response.status_code == 200:
            result = response.json()
            if result.get("success"):
                return result["data"]["formattedJson"]
        
        raise Exception(f"API Error: {response.text}")
    
    def create_document(self, data, name=None):
        """Create a new document"""
        response = requests.post(f"{self.base_url}/documents", json={
            "data": data,
            "name": name
        })
        
        if response.status_code == 200:
            result = response.json()
            if result.get("success"):
                return result["data"]
        
        raise Exception(f"API Error: {response.text}")
    
    def manipulate_json(self, operation, json_data, **kwargs):
        """Perform JSON manipulation"""
        payload = {
            "operation": operation,
            "jsonData": json_data,
            **kwargs
        }
        
        response = requests.post(f"{self.base_url}/json/manipulate", json=payload)
        
        if response.status_code == 200:
            result = response.json()
            if result.get("success"):
                return result["data"]["result"]
        
        raise Exception(f"API Error: {response.text}")

def run_tests():
    """Run comprehensive test suite"""
    print("🚀 JSON Canvas AI - Python Client Test Suite\n")
    
    print("=" * 50)
    print("🔍 BASIC API TESTS")
    print("=" * 50 + "\n")
    
    # Test basic endpoints
    test_endpoint("API Root Documentation", f"{BASE_URL}")
    test_endpoint("Available AI Models", f"{BASE_URL}/models")
    
    print("=" * 50)
    print("🤖 AI FEATURES TESTS")
    print("=" * 50 + "\n")
    
    # Test AI features
    test_endpoint("AI Text to JSON Conversion", f"{BASE_URL}/ai/convert-text", "POST", {
        "rawText": "Product: iPhone 15\nPrice: $999\nStorage: 128GB, 256GB, 512GB\nColors: Blue, Pink, Black, White\nRating: 4.5/5",
        "instructions": "Create a detailed product object with arrays for storage and color options"
    })
    
    test_endpoint("AI Field Enhancement", f"{BASE_URL}/ai/enhance-field", "POST", {
        "fieldContent": "Good software",
        "userPrompt": "Rewrite as professional software product description with technical benefits"
    })
    
    test_endpoint("AI JSON Formatting", f"{BASE_URL}/ai/format-json", "POST", {
        "jsonString": '{name:"Alice",age:28,skills:["python","javascript","react",],active:true,}',
        "instructions": "Fix syntax errors and format with proper indentation"
    })
    
    print("=" * 50)
    print("📄 DOCUMENT MANAGEMENT TESTS")
    print("=" * 50 + "\n")
    
    # Test document creation
    doc_result = test_endpoint("Create Document", f"{BASE_URL}/documents", "POST", {
        "data": {
            "company": "TechCorp",
            "employees": 150,
            "departments": ["Engineering", "Sales", "Marketing", "HR"],
            "founded": 2020,
            "revenue": "$5.2M"
        },
        "name": "Company Profile"
    })
    
    print("=" * 50)
    print("🔧 JSON MANIPULATION TESTS")
    print("=" * 50 + "\n")
    
    # Test JSON manipulations
    test_data = {"user": {"name": "Alice", "age": 28}}
    
    test_endpoint("Add Property", f"{BASE_URL}/json/manipulate", "POST", {
        "operation": "addProperty",
        "jsonData": test_data,
        "path": ["user"],
        "key": "department",
        "value": "Engineering"
    })
    
    test_endpoint("Add Array Item", f"{BASE_URL}/json/manipulate", "POST", {
        "operation": "addItem",
        "jsonData": {"skills": ["Python", "JavaScript"]},
        "path": ["skills"],
        "value": "React"
    })
    
    test_endpoint("Validate JSON", f"{BASE_URL}/json/manipulate", "POST", {
        "operation": "validate",
        "jsonData": {"user": {"name": "Alice", "age": 28, "active": True}}
    })
    
    print("=" * 50)
    print("🎯 PYTHON CLIENT DEMO")
    print("=" * 50 + "\n")
    
    try:
        client = JSONCanvasClient()
        
        # Demo workflow
        print("🔄 Python Client Demo Workflow:")
        print("1. Converting product data to JSON...")
        
        product_text = """
        Smart Watch Pro
        Price: $299
        Features: Heart rate monitoring, GPS tracking, Water resistant, 7-day battery
        Colors: Black, Silver, Gold
        Sizes: 42mm, 46mm
        """
        
        product_json = client.convert_text_to_json(product_text, "Create structured product object")
        print(f"✅ Converted to JSON: {json.dumps(product_json, indent=2)}")
        
        print("\n2. Enhancing product description...")
        enhanced_desc = client.enhance_field(
            "Smart watch with health features",
            "Create compelling marketing copy highlighting health and fitness benefits"
        )
        print(f"✅ Enhanced description: {enhanced_desc}")
        
        print("\n3. Adding enhanced description to product...")
        final_product = client.manipulate_json(
            "addProperty",
            product_json,
            path=[],
            key="marketingDescription",
            value=enhanced_desc
        )
        print(f"✅ Final product: {json.dumps(final_product, indent=2)[:300]}...")
        
        print("\n4. Creating document...")
        doc = client.create_document(final_product, "Smart Watch Product Data")
        print(f"✅ Document created: {doc['name']} (ID: {doc['id']})")
        
    except Exception as e:
        print(f"❌ Python client demo failed: {e}")
    
    print("\n" + "=" * 50)
    print("✅ ALL TESTS COMPLETED!")
    print("=" * 50 + "\n")
    print("🎉 JSON Canvas AI Python client is working perfectly!")
    print("📚 Full API documentation: http://localhost:9002/api")

if __name__ == "__main__":
    try:
        run_tests()
    except KeyboardInterrupt:
        print("\n🛑 Tests interrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n❌ Test suite failed: {e}")
        sys.exit(1)

File: test-api.js
#!/usr/bin/env node

// Complete API Test Suite for JSON Canvas AI Headless API
// Run with: node test-api.js

const baseUrl = 'http://localhost:9002/api';

console.log('🚀 JSON Canvas AI - Headless API Test Suite\n');

async function testEndpoint(name, url, options = {}) {
  try {
    console.log(`📋 Testing: ${name}`);
    const response = await fetch(url, options);
    const data = await response.json();
    
    if (response.ok) {
      console.log(`✅ ${name} - SUCCESS`);
      if (data.data) {
        console.log(`📄 Result:`, JSON.stringify(data.data, null, 2).slice(0, 200) + '...');
      } else {
        console.log(`📄 Response:`, JSON.stringify(data, null, 2).slice(0, 200) + '...');
      }
    } else {
      console.log(`❌ ${name} - FAILED`);
      console.log(`📄 Error:`, JSON.stringify(data, null, 2));
    }
    console.log('');
    return data;
  } catch (error) {
    console.log(`❌ ${name} - ERROR: ${error.message}`);
    console.log('');
    return null;
  }
}

async function runTests() {
  console.log('==================================================');
  console.log('🔍 BASIC API TESTS');
  console.log('==================================================\n');

  // Test 1: API Root
  await testEndpoint('API Root Documentation', `${baseUrl}`);

  // Test 2: Models List
  await testEndpoint('Available AI Models', `${baseUrl}/models`);

  console.log('==================================================');
  console.log('🤖 AI FEATURES TESTS');  
  console.log('==================================================\n');

  // Test 3: Convert Text to JSON
  const textConversionResult = await testEndpoint('AI Text to JSON Conversion', `${baseUrl}/ai/convert-text`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      rawText: 'John Doe, 30, Engineer, New York\nJane Smith, 25, Designer, Los Angeles\nBob Johnson, 35, Manager, Chicago',
      instructions: 'Create an array of employee objects with name, age, job, and city'
    })
  });

  // Test 4: Enhance Field
  await testEndpoint('AI Field Enhancement', `${baseUrl}/ai/enhance-field`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      fieldContent: 'Great product',
      userPrompt: 'Rewrite as a compelling e-commerce product description with features and benefits'
    })
  });

  // Test 5: Format JSON
  await testEndpoint('AI JSON Formatting', `${baseUrl}/ai/format-json`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      jsonString: '{name:"John",age:30,active:true,tags:["dev","manager",],}',
      instructions: 'Fix all syntax errors and format beautifully'
    })
  });

  console.log('==================================================');
  console.log('📄 DOCUMENT MANAGEMENT TESTS');
  console.log('==================================================\n');

  // Test 6: Create Document
  const documentResult = await testEndpoint('Create Document', `${baseUrl}/documents`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      data: {
        project: 'JSON Canvas Test',
        status: 'active',
        team: ['Alice', 'Bob', 'Charlie'],
        budget: 10000,
        deadline: '2024-12-31'
      },
      name: 'Project Alpha'
    })
  });

  // Test 7: Retrieve Document (if creation succeeded)
  if (documentResult && documentResult.data && documentResult.data.id) {
    await testEndpoint('Retrieve Document', `${baseUrl}/documents/${documentResult.data.id}`);
    
    // Test 8: Update Document
    await testEndpoint('Update Document', `${baseUrl}/documents/${documentResult.data.id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        data: {
          project: 'JSON Canvas Test - Updated',
          status: 'in-progress',
          team: ['Alice', 'Bob', 'Charlie', 'Diana'],
          budget: 15000,
          deadline: '2024-12-31',
          progress: '25%'
        },
        addToHistory: true
      })
    });
  }

  console.log('==================================================');
  console.log('🔧 JSON MANIPULATION TESTS');
  console.log('==================================================\n');

  // Test 9: Add Property
  await testEndpoint('Add Property to JSON', `${baseUrl}/json/manipulate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      operation: 'addProperty',
      jsonData: { user: { name: 'John', age: 30 } },
      path: ['user'],
      key: 'email',
      value: 'john@example.com'
    })
  });

  // Test 10: Add Array Item
  await testEndpoint('Add Item to Array', `${baseUrl}/json/manipulate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      operation: 'addItem',
      jsonData: { fruits: ['apple', 'banana'] },
      path: ['fruits'],
      value: 'orange'
    })
  });

  // Test 11: Set Value
  await testEndpoint('Set JSON Value', `${baseUrl}/json/manipulate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      operation: 'setValue',
      jsonData: { user: { name: 'John', age: 30 } },
      path: ['user', 'name'],
      value: 'John Smith'
    })
  });

  // Test 12: Rename Property
  await testEndpoint('Rename JSON Property', `${baseUrl}/json/manipulate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      operation: 'renameProperty',
      jsonData: { user: { name: 'John', age: 30 } },
      path: ['user'],
      key: 'name',
      newKey: 'fullName'
    })
  });

  // Test 13: Validate JSON
  await testEndpoint('Validate JSON Structure', `${baseUrl}/json/manipulate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      operation: 'validate',
      jsonData: { user: { name: 'John', age: 30, active: true, tags: ['developer', 'manager'] } }
    })
  });

  // Test 14: Delete Property
  await testEndpoint('Delete JSON Property', `${baseUrl}/json/manipulate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      operation: 'delete',
      jsonData: { user: { name: 'John', age: 30, temp: 'remove-me' } },
      path: ['user'],
      key: 'temp'
    })
  });

  console.log('==================================================');
  console.log('🎯 COMPLEX WORKFLOW TEST');
  console.log('==================================================\n');

  // Complex workflow: CSV -> JSON -> Enhance -> Document
  console.log('🔄 Running complex workflow...\n');

  // Step 1: Convert CSV to JSON
  const csvData = 'Name,Email,Department,Salary\nAlice Johnson,alice@company.com,Engineering,75000\nBob Smith,bob@company.com,Design,65000\nCharlie Brown,charlie@company.com,Marketing,60000';
  
  const csvResult = await testEndpoint('Step 1: CSV to JSON', `${baseUrl}/ai/convert-text`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      rawText: csvData,
      instructions: 'Convert to array of employee objects with proper data types'
    })
  });

  if (csvResult && csvResult.data) {
    // Step 2: Parse and enhance the result
    let employees;
    try {
      employees = JSON.parse(csvResult.data.generatedJson);
    } catch (e) {
      console.log('❌ Could not parse AI-generated JSON');
      return;
    }

    // Step 3: Add calculated fields
    const enhancedResult = await testEndpoint('Step 2: Add Calculated Fields', `${baseUrl}/json/manipulate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        operation: 'addProperty',
        jsonData: { employees: employees },
        path: [],
        key: 'totalBudget',
        value: employees.reduce((sum, emp) => sum + (emp.salary || emp.Salary || 0), 0)
      })
    });

    if (enhancedResult && enhancedResult.data) {
      // Step 4: Create document with the enhanced data
      await testEndpoint('Step 3: Create Employee Database Document', `${baseUrl}/documents`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          data: enhancedResult.data.result,
          name: 'Employee Database - Auto Generated'
        })
      });
    }
  }

  console.log('==================================================');
  console.log('✅ ALL TESTS COMPLETED!');
  console.log('==================================================\n');
  console.log('🎉 JSON Canvas AI Headless API is fully functional!');
  console.log('📚 Check the documentation at: http://localhost:9002/api');
  console.log('📁 Example files created: presidium.json, headless-api-examples.md');
}

// Run the tests
runTests().catch(console.error);